doc/source/rllib/env-to-module-connector.rst

```diff
@@ -0,0 +1,546 @@
+            return gym.spaces.Box(-100.0, 100.0, (in_obs_space.shape[0] + 3,), np.float32)
+
+        def preprocess(self, observation, episode):
+            # Extract the last 3 rewards from the ongoing episode using a python `slice` object.
+            # Alternatively, you can also pass in a list of indices, [-3, -2, -1].
+            past_3_rewards = episode.get_rewards(indices=slice(-3, None))
+
+            # Concatenate the rewards to the actual observation.
+            new_observation = np.concatenate([
+                observation, np.array(past_3_rewards, np.float32)
+            ])
+
+            # Return the new observation.
+            return new_observation
+
+
+Note that the preceding example should work without any further action required on your model,
+whether it's a custom one or a default one provided by RLlib, as long as the model determines its input layer's
+size through the observation space. The connector pipeline correctly captures the observation
+space changes, from the environment's 1D-Box to the reward-enhanced, larger 1D-Box and
+passes this new observation space to your RLModule's :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.setup`
+method.
+
+
+Preprocessing observations in multi-agent setups
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In multi-agent setups, you have two options for preprocessing your agents' individual observations
+through customizing your env-to-module pipeline:
+
+1) Agent-by-agent: Using the same API as in the previous examples (:py:class:`~ray.rllib.connectors.env_to_module.observation_preprocessor.SingleAgentObservationPreprocessor`),
+   you can apply a single preprocessing logic across all agents. However, in case you need one distinct preprocessing
+   logic per ``AgentID``, lookup the agent information from the provided ``episode`` argument in the
+   :py:meth:`~ray.rllib.connectors.env_to_module.observation_preprocessor.SingleAgentObservationPreprocessor.preprocess` method:
+
+   .. testcode::
+        :skipif: True
+
+        def recompute_output_observation_space(self, in_obs_space, in_act_space):
+            # `in_obs_space` is a `Dict` space, mapping agent IDs to individual agents' spaces.
+            # Alter this dict according to which agents you want to preprocess observations for
+            # and return the new `Dict` space.
+
+            # For example:
+            return gym.spaces.Dict({
+                "some_agent_id": [obs space],
+                "other_agent_id": [another obs space],
+                ...
+            })
+
+        def preprocess(self, observation, episode):
+
+            # Skip preprocessing for certain agent ID(s).
+            if episode.agent_id != "some_agent_id":
+                return observation
+
+            # Preprocess other agents' observations.
+            ...
+
+1) Multi-agent preprocessor with access to the entire multi-agent observation dict: Alternatively, you can subclass the
```

Comment:
I would emphasize the names of the different methods.

Reply:
not sure I understand the request.